### ISSUE TITLE
Angular 6: Can't resolve 'stream'

### ISSUE THREAD

**Issue by mditrolio** [DevExpress/devextreme-angular#776]
* Specify the version of the devextreme-angular and devextreme packages you are using.
    devexteme version: **17.2.7**
    devextreme-angular version: **17.2.7**

* Specify the type of the issue (check one with "x"):
[x] bug Report 
[ ] feature request 
[ ] support inquiry

We're trying to upgrade our project to Angular 6, but we see the following error when including any Dx* modules:

```
ERROR in ./node_modules/jszip/lib/readable-stream-browser.js
Module not found: Error: Can't resolve 'stream' in 'c:\Source\TestProject\node_modules\jszip\lib'
```

I've tried creating a clean Angular 6 project and just adding DevExtreme, but the results are the same.

**Comment by keserwan** [comment]
module.exports = require("readable-stream");
//module.exports = require("stream");

just replace this line of code in the file ("readable-stream-browser.js") until it is fixed by Angular/CLI

**Comment by keserwan** [comment]
I migrated today to angular 6 with devextreme 17.2.7
everything went well :)
and had to do this replace above.

**Comment by wartab** [comment]
When I try this solution, I get the following error:

```
ERROR in ./node_modules/devextreme/client_exporter/excel_creator.js
Module not found: Error: Can't resolve 'C:\Users\Vincent Lycoops\projects\wa\src\node_modules\jszip\dist\jszip.min.js' in 'C:\Users\Vincent Lycoops\projects\wa\node_modules\devextreme\client_exporter'
```

It's trying to include jszip from src/node_modules rather than node_modules.

**Comment by wartab** [comment]
That might fix it in devextreme's case, but I have other dependencies relying on correct behavior for jszip inclusion.

**Comment by jakehockey10** [comment]
I've followed the suggested fix, and I'm still getting the same issue.  The difference for me is that I am working with an Angular 6 workspace that has multiple applications in it.  I've made the `"paths": {...}` addition to one of the applications' `tsconfig.app.json` file, and I had to add two more `../` to the path, like this:

``` json
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "baseUrl": "./",
    "module": "es2015",
    "types": [],
    "paths": {
      "jszip": ["../../../node_modules/jszip/dist/jszip.min.js"]
    }
  },
  "exclude": ["test.ts", "**/*.spec.ts"]
}
```

But I still get the same error.  Is there anything I have to do differently when working in a workspace instead of a single application repo?

**Comment by GoshaFighten** [comment]
@jakehockey10 . I suggest you check what error message you are getting. For example, you may have your project in the `C:\MyProject` folder and the error says that `Can't resolve module 'C:\node_modules\jszip\dist\jszip.min.js' in ...`. This means that you need to remove one `../` from the path to `node_modules/jszip/dist/jszip.min.js`. If the error says that `Can't resolve module 'C:\MyProject\SomeOtherFolder\node_modules\jszip\dist\jszip.min.js' in ...`, this means you need to add one more `../`. And so on.

**Comment by jakehockey10** [comment]
@GoshaFighten , I tried a bunch of variations of this.  And I started to notice the point you made about the path in the error message.  But when I got the right amount of `../` in there to get the path correct in the error message, the error message was still there.  But then I saw that I added this to `tsconfig.app.json` instead of `tsconfig.json`.  Once I moved it to the root `tsconfig.json`, and adjusted the path to `node_modules` again, this seems to work now.  Thanks for taking a look!

**Comment by ssamayoa** [comment]
I just bumped with this problem with DX 18.1.3, this is what "ng version" reports:

**Comment by jakehockey10** [comment]
@ssamayoa, it was indeed the case for me that I had to add this fix in tsconfig.json and double check that my relative paths were correct. As mentioned already, the error message should give a pretty good clue as to what's wrong with the related be path if you scrutinize the path in the error message. You might see a folder missing or something that can help you figure out what your relative paths need to be.

**Comment by wartab** [comment]
I repeat: you should not have to play around with relative paths. This is a bug, the path in there is supposed to be relative to the folder that contains node_modules, not something random that depends on some magic.

**Comment by ssamayoa** [comment]
@wartab You mean a bug on DevExpress, algular-cli or in node?
Sorry if question sounds dumb but I'm not an expert on node.

**Comment by ardymalihi** [comment]
funny, I did npm install stream and worked for me :)

**Comment by jamesidw** [comment]
Can confirm... installing stream sorts this out. Thanks @ardymalihi

**Comment by jaypeelh** [comment]
installing stream worked.

**Comment by charce36** [comment]
Installing npm "stream" works for me.

`npm install stream --save`

**Comment by wartab** [comment]
Installing random packages that happen to have the name of a NodeJS API, but haven't been updated for 6 years... What a smart choice. Please request from Devextreme to patch this issue instead of "making it work" on your side with band-aids.

**Comment by jaykumar1210** [comment]
I did npm install stream and worked for me too

**Comment by PavelProg1984** [comment]
npm install stream --save помогло!

**Comment by carlosariash** [comment]
Installing npm "stream" works for me.

npm install stream --save

**Comment by hairangsua** [comment]
npm install stream --save works for me

**Comment by engineerSofware** [comment]
npm install stream --save works for me

**Comment by vecchiotom** [comment]
npm install stream --save works for me too

**Comment by briosheje** [comment]
Still experiencing the same issue with Angular 7, Devextreme ^18.2.3 and ionic 4.

**Comment by ahasan09** [comment]
npm install stream --save works for me too :)

**Comment by briosheje** [comment]
Is any fix for this issue even planned? It doesn't seems to be greatest choice ever to actually install a repository having more than 6 years of inactivity.

**Comment by cecilialimay** [comment]
npm install stream funciona para mi en angular 7

**Comment by tmtron** [comment]
so installing the `stream` package seems to be no good idea,
but registering JSZip in the tsconfig.json file (as suggested [here](https://github.com/DevExpress/devextreme-angular/issues/776#issuecomment-386706260)), does not work for me, because then the build fails with this:

```
Error:(9, 24) TS7016: Could not find a declaration file for module 'jszip'. 'ROOT/node_modules/jszip/dist/jszip.min.js' implicitly has an 'any' type.
```
see this [Stackoverflow question](https://stackoverflow.com/questions/56635317/typescript-does-not-find-jszip-types/) for details.

So does someone know a solution that fixes both issues?

**Comment by Gamester2097** [comment]
> module.exports = require("readable-stream");
> //module.exports = require("stream");
> 
> just replace this line of code in the file ("readable-stream-browser.js") until it is fixed by Angular/CLI

This worked for me

**Comment by Domvel** [comment]
The npm package `stream` is about 5..7 years old in version `0.0.2`. ... No, thanks. 
But for the first step, solved by adding the path in tsconfig.

**Comment by gabrielrb** [comment]
> 
> 
> ```
>     "paths": {
>       "jszip": [
>         "../node_modules/jszip/dist/jszip.min.js"
>       ]
>     }
> ```
> 
> put 2 dots before "node_modules"

thanks !!! this worked for me.

**Comment by HFTSOL** [comment]
Using version 19.1.6 and this is still broken.  Is there any intent to fix this issue?

**Comment by wartab** [comment]
There's no intent on their end. They say it's not caused by them and stopped replying since then. After this issue and several others, we have started migrating all our Devextreme components to mainly Material Design. The performance gains are gigantic, too.

**Comment by ssamayoa** [comment]
I no longer have this problem with Angular 8.x

**Comment by priyankawadekar** [comment]
> 
> 
> There's no intent on their end. They say it's not caused by them and stopped replying since then. After this issue and several others, we have started migrating all our Devextreme components to mainly Material Design. The performance gains are gigantic, too.

Can you please tell which material plugin is best to replace JSZip ?

**Comment by priyankawadekar** [comment]
I am still facing this issue in angular 8 - 
[INFO] ERROR in ./node_modules/jszip/lib/readable-stream-browser.js
[INFO] Hash: a6936540e76501d0d1ae
[INFO] Module not found: Error: Can't resolve 'stream' in 'D:\Workspace\node_modules\jszip\lib'

**Comment by MoAsmar** [comment]
it is a bug in "readable-stream" itself, it is missing package stream in the dependencies.

![image](https://user-images.githubusercontent.com/8542126/71416353-74c22400-2679-11ea-8566-5d3050d4e0e4.png)


you have to install stream 
`npm i stream`

**Comment by bvdwalt** [comment]
> it is a bug in "readable-stream" itself, it is missing package stream in the dependencies.
> 
> ![image](https://user-images.githubusercontent.com/8542126/71416353-74c22400-2679-11ea-8566-5d3050d4e0e4.png)
> 
> you have to install stream
> `npm i stream`

Thanks man, sorted it out for me.

**Comment by secret-dave** [comment]
The guide to adding DevExtreme to an Angular application has this step defined: [JSZip Registration](https://js.devexpress.com/DevExtreme/Guide/Angular_Components/Getting_Started/Add_DevExtreme_to_an_Angular_CLI_Application/#Register_3rd-Party_Dependencies_in_Angular_CLI_6+/JSZip_Registration), and it fixes the issue.

**Comment by tmtron** [comment]
For me the issue is now solved in jszip version 3.3.0: see [Fix browser module resolution #614](https://github.com/Stuk/jszip/pull/614)

I don't need the outdated [stream package](https://www.npmjs.com/package/stream) or use the [JSZip Registration](https://js.devexpress.com/DevExtreme/Guide/Angular_Components/Getting_Started/Add_DevExtreme_to_an_Angular_CLI_Application/#Register_3rd-Party_Dependencies_in_Angular_CLI_6+/JSZip_Registration) workaround (which broke my typescript type resolution: see [SO typescript does not find jszip types](https://stackoverflow.com/questions/56635317/typescript-does-not-find-jszip-types/))
